src/model/history.js

```diff
@@ -36,109 +35,176 @@ export default class History {
-		if ( delta && this._deltas[ this._deltas.length - 1 ] !== delta ) {
-			const index = this._deltas.length;
+	*getDeltas( from = 0, to = Number.POSITIVE_INFINITY ) {
+		// No deltas added, nothing to yield.
+		if ( this._deltas.length === 0 ) {
+			return;
+		}
 
-			this._deltas[ index ] = delta;
-			this._historyPoints.set( delta.baseVersion, index );
+		// Will throw if base version is incorrect.
+		let fromIndex = this._getIndex( from );
+
+		// Base version is too low or too high and is not found in history.
+		if ( fromIndex == -1 ) {
+			return;
+		}
+
+		// We have correct `fromIndex` so let's iterate starting from it.
+		while ( fromIndex < this._deltas.length ) {
+			const delta = this._deltas[ fromIndex++ ];
+
+			if ( delta.baseVersion >= to ) {
+				break;
+			}
+
+			yield delta;
 		}
 	}
 
 	/**
-	 * Transforms out-dated delta by all deltas that were added to the history since the given delta's base version. In other
-	 * words, it makes the delta up-to-date with the history. The transformed delta(s) is (are) ready to be applied
-	 * to the {@link engine.model.Document document}.
+	 * Returns one or more deltas from history that bases on given `baseVersion`. Most often it will be just
+	 * one delta, but if that delta got updated by multiple deltas, all of those updated deltas will be returned.
 	 *
-	 * @param {engine.model.delta.Delta} delta Delta to update.
-	 * @returns {Array.<engine.model.delta.Delta>} Result of transformation which is an array containing one or more deltas.
+	 * @see engine.model.History#updateDelta
+	 * @param {Number} baseVersion Base version of the delta to retrieve.
+	 * @returns {Array.<engine.model.delta.Delta>|null} Delta with given base version or null if no such delta is in history.
 	 */
-	getTransformedDelta( delta ) {
-		if ( delta.baseVersion === this._nextHistoryPoint ) {
-			return [ delta ];
+	getDelta( baseVersion ) {
+		let index = this._historyPoints.get( baseVersion );
+
+		if ( index === undefined ) {
+			return null;
 		}
 
-		let transformed = [ delta ];
+		const deltas = [];
 
-		for ( let historyDelta of this.getDeltas( delta.baseVersion ) ) {
-			let allResults = [];
+		for ( index; index < this._deltas.length; index++ ) {
+			const delta = this._deltas[ index ];
 
-			for ( let deltaToTransform of transformed ) {
-				const transformedDelta = History._transform( deltaToTransform, historyDelta );
-				allResults = allResults.concat( transformedDelta );
+			if ( delta.baseVersion != baseVersion ) {
+				break;
 			}
 
-			transformed = allResults;
+			deltas.push( delta );
 		}
 
-		// Fix base versions.
-		let baseVersion = transformed[ 0 ].operations[ 0 ].baseVersion;
-
-		for ( let i = 0; i < transformed.length; i++ ) {
-			transformed[ i ].baseVersion = baseVersion;
-			baseVersion += transformed[ i ].operations.length;
-		}
+		return deltas.length === 0 ? null : deltas;
+	}
```

Comment:
Thanks for the analysis. The remaining question is what should happen if one user undo. That reversed delta propagate, but how it should be applied. If it will be applied as a regular data we will have similar situation we had with undo v.1. It does not necessarily need to be an issue, since you can not undo other users changes. Alternatively we could compress deltas there too, but then we need to move the mechanism from the undo feature to the engine. Anyway, there are solution, what is enough for now.